ql/src/test/org/apache/hadoop/hive/ql/optimizer/calcite/stats/TestFilterSelectivityEstimator.java

```diff
@@ -159,7 +159,7 @@ public void testIsHistogramAvailableWhenEmptyArray() {
 
   @Test
   public void testLessThanSelectivity() {
-    Assert.assertEquals(0.6153846153846154, lessThanSelectivity(KLL, 3), DELTA);
```

Comment:
I am still not convinced that those values should change, from what I understood @okumin found a way to have the same results? Have you applied his suggestions on the way we compute and query the data sketch?

Reply:
Fixed in https://github.com/apache/hive/pull/5444/commits/ab110b6d7ce29b2a0bc916cf3dd62634fdf6df23

I think all histogram stats related tests are changed back now. :)

Reply:
But i still can not figure why skecth functions Qtests changed. Such as `sketches_materialized_view_cume_dist.q.out`.

I noticed that https://github.com/apache/hive/pull/1126#discussion_r442360252 also did similar change. 
Maybe @kgyrtkirk you can give some thread? Thanks.

Reply:
I have found the root cause: some skecth functions have multiple `evaluate` method, but in Hive we always get the first `evaluate` method, see:
https://github.com/apache/hive/blob/548990dfad78b3d89a334c875d8a6708ef475e88/ql/src/java/org/apache/hadoop/hive/ql/exec/DataSketchesFunctions.java#L288-L297

Here i gave a example: In this commit https://github.com/apache/datasketches-hive/commit/b6c4d01ff9539d2aff520569277c14d479695bb0#diff-722be040808b3a7ef18a0e86ced7e686fcf553002f82d8332d02d29f28b265a9 , it added another `evaluate `method with the new added param `QuantileSearchCriteria.INCLUSIVE`, and then, some Qtest which uses `cdf` will change its result.

And some other skecth functions may have more `evaluate ` method, we also choose the first one.

IMO, we can not find a good way to choose the best `evaluate` method, what i can do is regenerate the skecth functions Qtests outout. :(

Do you have any other good ideas? @asolimando

Reply:
How about modifying datasketches-java so that it is consistent with the previous version?

Reply:
@okumin, if they are open to that change, we could definitely try and that would solve our problems. But since the behavior changed in a major release, technically it's something that could be intended and expected, although it makes our life more complicated.

@zhangbutao sorry I missed your previous message! What about moving the evaluate method into a map which is filled when invoking `register` for each skect function we support? It's less elegant maybe than the existing, but we would regain full control on what we invoke at eval time and we are anyway already specifying many info in the register function, so it does seem to be coherent with the current design.

What do you guys think?

Reply:
> How about modifying datasketches-java so that it is consistent with the previous version?

Sorry, I intended modifying **datasketches-hive**, not datasketches-java. The Hive community may have some control over the library.

Reply:
> What about moving the evaluate method into a map which is filled when invoking register for each skect function we support?

@asolimando But we don't know what `evaluate `methods of sketch funtion we should support. In fact, we can run all `evaluate` methods of each function, but we just pick the first one & we can not determine which `evaluate ` method is better or is what we need.

> The Hive community may have some control over the library datasketches-hive.

@okumin I think this is too hard :(. I don't think we can do anything about the datasketches. Unless we backport the whole base code of datasketches-hive into Hive repo, but this may cause some codes inconsistency&conflict with datasketches in the futrue.

Reply:
Just let me clarify the problem. In the case of `GetCdfUDF`, originally, we provided `(ByteWritable, Float...) -> List<Double>`. Datasketches Hive 2.0 provides `(ByteWritable, Float...) -> List<Double>` and `(ByteWritable, Boolean, Float...) -> List<Double>`, and the first one is now an alias of `(ByteWritable, true, Float...) -> List<Double>`. While using Datasketches Hive 1.2, the method with 2-arity was working as `(ByteWritable, false, Float...) -> List<Double>`. So, the original query unit test failed as the 2-arity method started using the INCLUSIVE mode.

One more, I guess we don't have to keep 100% compatibility of sketch UDFs because it returns an approximate result. However, we will encounter incompatibilities when the sketches, not the final result, are directly stored as binary. Information stored in HMS is the primary example. Do we have another case to directly store sketches on the system side? e.g. Materialized View

Reply:
> I have found the root cause: some skecth functions have multiple `evaluate` method, but in Hive we always get the first `evaluate` method, see:
> 
> https://github.com/apache/hive/blob/548990dfad78b3d89a334c875d8a6708ef475e88/ql/src/java/org/apache/hadoop/hive/ql/exec/DataSketchesFunctions.java#L288-L297
> 
> Here i gave a example: In this commit [apache/datasketches-hive@b6c4d01#diff-722be040808b3a7ef18a0e86ced7e686fcf553002f82d8332d02d29f28b265a9](https://github.com/apache/datasketches-hive/commit/b6c4d01ff9539d2aff520569277c14d479695bb0#diff-722be040808b3a7ef18a0e86ced7e686fcf553002f82d8332d02d29f28b265a9) , it added another `evaluate `method with the new added param `QuantileSearchCriteria.INCLUSIVE`, and then, some Qtest which uses `cdf` will change its result.
> 
> And some other skecth functions may have more `evaluate ` method, we also choose the first one.
> 
> IMO, we can not find a good way to choose the best `evaluate` method, what i can do is regenerate the skecth functions Qtests outout. :(
> 
> Do you have any other good ideas? @asolimando

Actually the method you cite only infers the return type, it's not exactly what we want, although the root cause it's something similar I guess, someplace else.

I have traced some functions finding by reflection the correct `evaluate` method for UDF and UADF, respectively https://github.com/apache/hive/blob/master/udf/src/java/org/apache/hadoop/hive/ql/exec/UDFMethodResolver.java#L49
and https://github.com/apache/hive/blob/master/udf/src/java/org/apache/hadoop/hive/ql/exec/DefaultUDAFEvaluatorResolver.java#L57 

The problem is not in those methods, as they are receiving the actual function inputs and just matching on them, what we would need to do, is to intercept where this UDF/UDAF are written in the AST/plan and replace `$udf($binary, $float)` with `$udf($binary, false, $float)` so that the right evaluate function is then invoked.

I tried to track down places where the sketches are introduced (e.g., rules like https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRewriteToDataSketchesRules.java) but without much luck so far.

Reply:
> what we would need to do, is to intercept where this UDF/UDAF are written in the AST/plan and replace $udf($binary, $float) with $udf($binary, false, $float) so that the right evaluate function is then invoked.

This seems like a good idea. :)   I will try to explore this way, too.

Because incompatible datasketches changes lead to a lot of work, it seems difficult to control datasketches-hive.  I hope this PR is the last one to do this work.  :(